# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [1]:
import pandas as pd
from IPython.display import display
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.metrics import (
    r2_score, mean_squared_error, mean_absolute_error,
    accuracy_score, balanced_accuracy_score, precision_score,
    recall_score, f1_score, confusion_matrix,
)

## Regression Model Evaluation

In [2]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
data = pd.read_csv('housing.csv', header=None, delimiter=r"\s+", names=column_names)

In [3]:
"""
CRIM - per capita crime rate by town
ZN - proportion of residential land zoned for lots over 25,000 sq.ft.
INDUS - proportion of non-retail business acres per town.
CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)
NOX - nitric oxides concentration (parts per 10 million)
RM - average number of rooms per dwelling
AGE - proportion of owner-occupied units built prior to 1940
DIS - weighted distances to five Boston employment centres
RAD - index of accessibility to radial highways
TAX - full-value property-tax rate per $10,000
PTRATIO - pupil-teacher ratio by town
B - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
LSTAT - % lower status of the population
MEDV - Median value of owner-occupied homes in $1000's"""

Out[3]: "\nCRIM - per capita crime rate by town\nZN - proportion of residential land zoned for lots over 25,000 sq.ft.\nINDUS - proportion of non-retail business acres per town.\nCHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)\nNOX - nitric oxides concentration (parts per 10 million)\nRM - average number of rooms per dwelling\nAGE - proportion of owner-occupied units built prior to 1940\nDIS - weighted distances to five Boston employment centres\nRAD - index of accessibility to radial highways\nTAX - full-value property-tax rate per $10,000\nPTRATIO - pupil-teacher ratio by town\nB - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town\nLSTAT - % lower status of the population\nMEDV - Median value of owner-occupied homes in $1000's"


In [4]:
data

Out[4]: 
        CRIM    ZN  INDUS  CHAS    NOX  ...    TAX  PTRATIO       B  LSTAT  MEDV
0    0.00632  18.0   2.31     0  0.538  ...  296.0     15.3  396.90   4.98  24.0
1    0.02731   0.0   7.07     0  0.469  ...  242.0     17.8  396.90   9.14  21.6
2    0.02729   0.0   7.07     0  0.469  ...  242.0     17.8  392.83   4.03  34.7
3    0.03237   0.0   2.18     0  0.458  ...  222.0     18.7  394.63   2.94  33.4
4    0.06905   0.0   2.18     0  0.458  ...  222.0     18.7  396.90   5.33  36.2
..       ...   ...    ...   ...    ...  ...    ...      ...     ...    ...   ...
501  0.06263   0.0  11.93     0  0.573  ...  273.0     21.0  391.99   9.67  22.4
502  0.04527   0.0  11.93     0  0.573  ...  273.0     21.0  396.90   9.08  20.6
503  0.06076   0.0  11.93     0  0.573  ...  273.0     21.0  396.90   5.64  23.9
504  0.10959   0.0  11.93     0  0.573  ...  273.0     21.0  393.45   6.48  22.0
505  0.04741   0.0  11.93     0  0.573  ...  273.0     21.0  396.90   7.88  11.9

[506 rows x 14 col

## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [5]:
X_reg = data.drop(columns="MEDV")
y_reg = data["MEDV"]
X_reg_train, X_reg_test, y_reg_train, y_reg_test = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)
print(f"Training samples: {len(X_reg_train)}; testing samples: {len(X_reg_test)}")

Training samples: 404; testing samples: 102


## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [6]:
linear_model = LinearRegression()
linear_model.fit(X_reg_train, y_reg_train)
y_reg_train_pred = linear_model.predict(X_reg_train)
y_reg_test_pred = linear_model.predict(X_reg_test)

/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279:

## 3. Calculate and print R-squared for both the training and the testing set.

In [7]:
print(f"Training R-squared: {r2_score(y_reg_train, y_reg_train_pred):.4f}")
print(f"Testing R-squared: {r2_score(y_reg_test, y_reg_test_pred):.4f}")

Training R-squared: 0.7509
Testing R-squared: 0.6688


## 4. Calculate and print mean squared error for both the training and the testing set.

In [8]:
# MSE is measured in squared thousands of dollars.
print(f"Training MSE: {mean_squared_error(y_reg_train, y_reg_train_pred):.4f}")
print(f"Testing MSE: {mean_squared_error(y_reg_test, y_reg_test_pred):.4f}")

Training MSE: 21.6414
Testing MSE: 24.2911


## 5. Calculate and print mean absolute error for both the training and the testing set.

In [9]:
# MAE is measured in thousands of dollars.
print(f"Training MAE: {mean_absolute_error(y_reg_train, y_reg_train_pred):.4f}")
print(f"Testing MAE: {mean_absolute_error(y_reg_test, y_reg_test_pred):.4f}")

Training MAE: 3.3148
Testing MAE: 3.1891


## Classification Model Evaluation

In [10]:
from sklearn.datasets import load_iris
data = load_iris()

In [11]:
print(data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

In [12]:
column_names = data.feature_names

In [13]:
df = pd.DataFrame(data['data'],columns=column_names)

In [14]:
df

Out[14]: 
     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


In [15]:
target = pd.DataFrame(data.target)

In [16]:
data.keys()

Out[16]: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [17]:
data['target_names']

Out[17]: array(['setosa', 'versicolor', 'virginica'], dtype='<U10')


## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [18]:
# The supplied target contains the flower class encoded as 0, 1, or 2.
y_cls = target.iloc[:, 0].rename("class")
X_cls_train, X_cls_test, y_cls_train, y_cls_test = train_test_split(
    df, y_cls, test_size=0.2, random_state=42, stratify=y_cls
)
print(f"Training samples: {len(X_cls_train)}; testing samples: {len(X_cls_test)}")
print("Test class counts:")
print(y_cls_test.value_counts().sort_index())

Training samples: 120; testing samples: 30
Test class counts:
class
0    10
1    10
2    10
Name: count, dtype: int64


## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [19]:
logistic_model = LogisticRegression(max_iter=1000, random_state=42)
logistic_model.fit(X_cls_train, y_cls_train)
y_cls_train_pred = logistic_model.predict(X_cls_train)
y_cls_test_pred = logistic_model.predict(X_cls_test)

## 8. Calculate and print the accuracy score for both the training and the testing set.

In [20]:
print(f"Training accuracy: {accuracy_score(y_cls_train, y_cls_train_pred):.4f}")
print(f"Testing accuracy: {accuracy_score(y_cls_test, y_cls_test_pred):.4f}")

Training accuracy: 0.9750
Testing accuracy: 0.9667


## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [21]:
print(f"Training balanced accuracy: {balanced_accuracy_score(y_cls_train, y_cls_train_pred):.4f}")
print(f"Testing balanced accuracy: {balanced_accuracy_score(y_cls_test, y_cls_test_pred):.4f}")

Training balanced accuracy: 0.9750
Testing balanced accuracy: 0.9667


## 10. Calculate and print the precision score for both the training and the testing set.

In [22]:
# Macro averaging gives each of the three flower classes equal weight.
print(f"Training macro precision: {precision_score(y_cls_train, y_cls_train_pred, average='macro', zero_division=0):.4f}")
print(f"Testing macro precision: {precision_score(y_cls_test, y_cls_test_pred, average='macro', zero_division=0):.4f}")

Training macro precision: 0.9752
Testing macro precision: 0.9697


## 11. Calculate and print the recall score for both the training and the testing set.

In [23]:
# Macro averaging gives each of the three flower classes equal weight.
print(f"Training macro recall: {recall_score(y_cls_train, y_cls_train_pred, average='macro', zero_division=0):.4f}")
print(f"Testing macro recall: {recall_score(y_cls_test, y_cls_test_pred, average='macro', zero_division=0):.4f}")

Training macro recall: 0.9750
Testing macro recall: 0.9667


## 12. Calculate and print the F1 score for both the training and the testing set.

In [24]:
# Macro averaging gives each of the three flower classes equal weight.
print(f"Training macro f1: {f1_score(y_cls_train, y_cls_train_pred, average='macro', zero_division=0):.4f}")
print(f"Testing macro f1: {f1_score(y_cls_test, y_cls_test_pred, average='macro', zero_division=0):.4f}")

Training macro f1: 0.9750
Testing macro f1: 0.9666


## 13. Generate confusion matrices for both the training and the testing set.

In [25]:
# Rows are true classes; columns are predicted classes.
class_names = list(data.target_names)
for split, actual, predicted in [
    ("Training", y_cls_train, y_cls_train_pred),
    ("Testing", y_cls_test, y_cls_test_pred),
]:
    print(f"{split} confusion matrix (rows=true, columns=predicted):")
    display(pd.DataFrame(
        confusion_matrix(actual, predicted, labels=[0, 1, 2]),
        index=pd.Index(class_names, name="True class"),
        columns=pd.Index(class_names, name="Predicted class"),
    ))

Training confusion matrix (rows=true, columns=predicted):
Testing confusion matrix (rows=true, columns=predicted):


Predicted class,setosa,versicolor,virginica
True class,,,
setosa,40,0,0
versicolor,0,38,2
virginica,0,1,39


Predicted class,setosa,versicolor,virginica
True class,,,
setosa,10,0,0
versicolor,0,9,1
virginica,0,0,10


## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

In [26]:
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier

# Compare models on exactly the same splits as the earlier exercises.
reg_models = {
    "Linear regression": linear_model,
    "Decision tree": DecisionTreeRegressor(random_state=42),
    "Random forest": RandomForestRegressor(n_estimators=200, random_state=42),
}
reg_rows = []
for name, model in reg_models.items():
    model.fit(X_reg_train, y_reg_train)
    for split, X, y in [("Train", X_reg_train, y_reg_train), ("Test", X_reg_test, y_reg_test)]:
        pred = model.predict(X)
        reg_rows.append({"Model": name, "Split": split, "R-squared": r2_score(y, pred),
                         "MSE": mean_squared_error(y, pred), "MAE": mean_absolute_error(y, pred)})
reg_comparison = pd.DataFrame(reg_rows)
display(reg_comparison.round(4))

cls_models = {
    "Logistic regression": logistic_model,
    "Decision tree": DecisionTreeClassifier(random_state=42),
    "Random forest": RandomForestClassifier(n_estimators=200, random_state=42),
}
cls_rows = []
for name, model in cls_models.items():
    model.fit(X_cls_train, y_cls_train)
    for split, X, y in [("Train", X_cls_train, y_cls_train), ("Test", X_cls_test, y_cls_test)]:
        pred = model.predict(X)
        cls_rows.append({"Model": name, "Split": split, "Accuracy": accuracy_score(y, pred),
                         "Balanced accuracy": balanced_accuracy_score(y, pred),
                         "Precision (macro)": precision_score(y, pred, average="macro", zero_division=0),
                         "Recall (macro)": recall_score(y, pred, average="macro", zero_division=0),
                         "F1 (macro)": f1_score(y, pred, average="macro", zero_division=0)})
        print(f"{name} — {split} confusion matrix:")
        display(pd.DataFrame(confusion_matrix(y, pred, labels=[0, 1, 2]),
                             index=class_names, columns=class_names))
cls_comparison = pd.DataFrame(cls_rows)
display(cls_comparison.round(4))

reg_test_scores = reg_comparison.query("Split == 'Test'")
cls_test_scores = cls_comparison.query("Split == 'Test'")
print("Lowest test MSE:", reg_test_scores.loc[reg_test_scores["MSE"].idxmin(), "Model"])
best_f1 = cls_test_scores["F1 (macro)"].max()
print("Highest test macro F1:", ", ".join(cls_test_scores.loc[
    cls_test_scores["F1 (macro)"].eq(best_f1), "Model"]))

Logistic regression — Train confusion matrix:
Logistic regression — Test confusion matrix:
Decision tree — Train confusion matrix:
Decision tree — Test confusion matrix:
Random forest — Train confusion matrix:
Random forest — Test confusion matrix:
Lowest test MSE: Random forest
Highest test macro F1: Logistic regression


/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/mithila/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279:

,Model,Split,R-squared,MSE,MAE
0,Linear regression,Train,0.7509,21.6414,3.3148
1,Linear regression,Test,0.6688,24.2911,3.1891
2,Decision tree,Train,1.0000,0.0000,0.0000
3,Decision tree,Test,0.8481,11.1428,2.5284
4,Random forest,Train,0.9776,1.9422,0.8973
5,Random forest,Test,0.8819,8.6580,2.0536


,setosa,versicolor,virginica
setosa,40,0,0
versicolor,0,38,2
virginica,0,1,39


,setosa,versicolor,virginica
setosa,10,0,0
versicolor,0,9,1
virginica,0,0,10


,setosa,versicolor,virginica
setosa,40,0,0
versicolor,0,40,0
virginica,0,0,40


,setosa,versicolor,virginica
setosa,10,0,0
versicolor,0,9,1
virginica,0,1,9


,setosa,versicolor,virginica
setosa,40,0,0
versicolor,0,40,0
virginica,0,0,40


,setosa,versicolor,virginica
setosa,10,0,0
versicolor,0,9,1
virginica,0,2,8


,Model,Split,Accuracy,Balanced accuracy,Precision (macro),Recall (macro),F1 (macro)
0,Logistic regression,Train,0.9750,0.9750,0.9752,0.9750,0.9750
1,Logistic regression,Test,0.9667,0.9667,0.9697,0.9667,0.9666
2,Decision tree,Train,1.0000,1.0000,1.0000,1.0000,1.0000
3,Decision tree,Test,0.9333,0.9333,0.9333,0.9333,0.9333
4,Random forest,Train,1.0000,1.0000,1.0000,1.0000,1.0000
5,Random forest,Test,0.9000,0.9000,0.9024,0.9000,0.8997


### Interpreting the comparison
Higher R-squared and classification scores are better; lower MSE and MAE are better. Compare test performance to assess generalization. A large train–test gap, especially a perfect training score for an unrestricted tree, suggests overfitting. Macro scores weight each Iris class equally; balanced accuracy is the mean per-class recall.

The rankings above describe this fixed split only. With just 30 Iris test examples, one error changes accuracy by about 3.3 percentage points. For model selection, use cross-validation on the training data and reserve a fresh holdout for final evaluation.

On this split, Random forest has the lowest regression test MSE. The displayed classification ranking includes all models tied for the highest test macro F1.